core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -777,6 +777,12 @@ class GroupCoordinator(val brokerId: Int,
 
     maybePrepareRebalance(group, s"Adding new member $memberId")
     group.removePendingMember(memberId)
+
+    // attempt to complete JoinGroup
+    if (group.hasAllMembersJoined) {
```

Comment:
I would suggest something like this:
```scala
        if (group.is(PreparingRebalance))
          joinPurgatory.checkAndComplete(GroupKey(group.groupId))
```
We already have this logic inside `doJoinGroup` down one path. Maybe we can pull the check into a common path?

There are a couple other things we need to fix:

1. Inside `tryCompleteHeartbeat`, we need to call `forceComplete` if the pending member has joined.
2. Inside `onExpireHeartbeat`, we need to check for rebalance completion after removal of the pending member. Basically we need logic similar to `removeMemberAndUpdateGroup`.

Reply:
+1 for using joinPurgatory.